core/dataset.cpp

```diff
@@ -784,9 +836,7 @@ std::ostream &operator<<(std::ostream &os, const Dim dim) {
 }
 
 Dataset operator+(const Dataset &lhs, const Dataset &rhs) {
-  Dataset res(lhs);
-  apply(plus_equals, res, rhs);
-  return res;
+  return apply_with_broadcast(plus, lhs, rhs);
 }
 
 Dataset operator+(const Dataset &lhs, const DatasetConstProxy &rhs) {
```

Comment:
All non-in-place overloads like this and others below must use `apply_with_broadcast`. You will also need a broadcasting version "with delay".

Reply:
I don't think a "with delay" version of `apply_with_broadcast` is needed given that it is always creating a new `Dataset`?

Reply:
Good point, provided that you remove all overloads with rvalue references.